### PR TITLE
Do not overwrite TMPDIR

### LINKED ISSUE #13325: Npm overrides TMPDIR for root users
#### I'm opening this issue because:
- [ ] npm is crashing.
- [ ] npm is producing an incorrect install.
- [x] npm is doing something I don't understand.
- [ ] Other (_see below for feature requests_):
#### What's going wrong?

npm assumes that if `unsafe-perm` is false (the default for root users) that we want to use the current working directory as `TMPDIR`.

https://github.com/npm/npm/blob/6da8d1d47eb3f933fffd47952079dd3d94e9fe25/lib/utils/lifecycle.js#L80

This leads to some surprising and annoying behavior when running npm scripts as root on docker.

This was originally reported here: https://github.com/npm/npm/issues/4531 but closed.

It seems that the original "fix" was presumed that you could not write to `/tmp` if `unsafe-perm` is unset. Given that this is the default for the `root` user, this doesn't make much sense to me.

Perhaps npm should check to see if it can write to the directory before silently overriding an environment variable?
#### How can the CLI team reproduce the problem?

Create a `package.json` with: 

``` json
...
"scripts": {
    "test": "echo $TMPDIR"
  },
...
```

Run `sudo npm test`
### supporting information:
- `npm -v` prints: 3.8.9
- `node -v` prints: v6.2.0
- `npm config get registry` prints:
  
  ```
  # npm config get registry
  npm info it worked if it ends with ok
  npm info using npm@3.8.9
  npm info using node@v6.2.0
  https://registry.npmjs.org/
  npm info ok
  ```
- Windows, OS X, or Linux?: Linux and OS X
- Network issues:
  - Geographic location where npm was run:
  - [ ] I use a proxy to connect to the npm registry.
  - [ ] I use a proxy to connect to the web.
  - [ ] I use a proxy when downloading Git repos.
  - [ ] I access the npm registry via a VPN
  - [ ] I don't use a proxy, but have limited or unreliable internet access.
- Container:
  - [ ] I develop using Vagrant on Windows.
  - [ ] I develop using Vagrant on OS X or Linux.
  - [x] I develop / deploy using Docker.
  - [ ] I deploy to a PaaS (Triton, Heroku).

### PR DESCRIPTION
**Pull request by aaronjensen**
Fixes #13325 

Maybe a check for 'nobody' would be more appropriate, I don't know, but this existing behavior is surprising and causes mysterious/confusing problems.

### CODE REVIEW

**Comment by othiym23** [comment]
As the comment indicates, just straight-up deleting those lines is going to cause problems for users where bash can't write to `$TMPDIR` as `nobody`. Having some kind of tests as to whether the CLI can write to the current directory and / or `$TMPDIR` and choosing one or the other (and maybe bailing out if there's _nowwhere_ we can write to) seems much less likely to cause unexpected regressions. Also, this needs a test to be merged. Thanks for the work you've done thus far!

**Comment by othiym23** [comment]
See also d7dcacea6f17437fd88321987ca6aa8fe3c017d8, when this behavior was added: `/bin/sh` on Solaris / SmartOS systems, which is "pure" Bourne shell, freaks out when it can't write to TMPDIR. Cross-platform compatibility! 💫

**Comment by aaronjensen** [comment]
@othiym23 Thank you for the feedback. Please see the updated commit and let me know if there's anything I can do to improve it.

File: lib/utils/lifecycle.js [hunk 1]
```diff
@@ -75,18 +76,38 @@ function lifecycle (pkg, stage, wd, unsafe, failOk, cb) {
...
+  cb(new Error('Cannot find writable TMPDIR, please set TMPDIR to something writable'))
+}
+
+function _isWritable (dir) {
+  try {
+    fs.accessSync(dir, fs.W_OK)
+    return true
+  } catch (e) {
+    return false
+  }
```
Comment (aaronjensen) [inline lib/utils/lifecycle.js]: I'm not a fan of using exceptions in this way, but it felt like an awkward place to have to use callbacks. If you'd prefer that let me know and I can switch it out.

File: test/tap/lifecycle.js [hunk 2]
```diff
@@ -42,3 +42,30 @@ test('lifecycle : rejects wd for ', function (t) {
     t.end()
   })
 })
+
+test('lifecycle : accepts TMPDIR that is writable', function (t) {
```
Comment (aaronjensen) [inline test/tap/lifecycle.js]: These tests assume TMPDIR is set properly and writeable, hopefully that's not a problem anywhere.

File: test/tap/lifecycle.js [hunk 3]
```diff
@@ -42,3 +42,30 @@ test('lifecycle : rejects wd for ', function (t) {
...
+    t.equal(tmpdir, wd)
+    t.end()
+  })
+})
+
+test('lifecycle : errors if neither TMPDIR or wd are writable', function (t) {
+  var wd = '/probably/not/writable/either'
+
+  lifecycle._getTemporaryDirectory(wd, {TMPDIR: '/probably/not/writable'}, function (er, tmpdir) {
+    t.type(er, Error)
```
Comment (aaronjensen) [inline test/tap/lifecycle.js]: I'm new to tap and `t.error` didn't seem to do what I expected. Is this the proper way to test this?

Comment (iarna) [inline test/tap/lifecycle.js]: Yeah `t.ifError` (and `t.error`) go against the `tap` grain and assert there WASN'T an error.

I would just do: `t.ok(er, 'assertion description')` to assert that an error _was_ produced.

File: lib/utils/lifecycle.js (hunk 1 above)
Comment (iarna) [inline lib/utils/lifecycle.js]: In a test, this is fine (I do it myself).

**Comment by aaronjensen** [comment]
Thanks @iarna. I updated the test to use `t.ok`. Is there anything else you'd like me to update on this or is it good to go?

**Review by iarna — APPROVED** [review]
LGTM

**Comment by iarna** [comment]
This does not pass on 0.10 because it uses `fs.accessSync` which is not available on 0.10. We've gotta support 0.10 a little longer so I'm going to defer merging it.

**Review by iarna — CHANGES_REQUESTED** [review]
There's an issue with this: It checks to see if `TMPDIR` is writable _by the current user_. But the error condition we're trying to guard against is `TMPDIR` not being writable by the uid & gid passed in.

I'm not sure there's a meaningful way to check for that, short of spawning a sub process to check for us.

**Comment by iarna** [comment]
As I ultimately found a problem with this that I don't know how to resolve, and there's been no reply in over six months, I'm going to close this. Please feel free to open a new PR if you'd like to continue this discussion. Thanks!

**Pull request closed without merging**